### PR TITLE
Introduce procModifier for parabolic magnetic field in track building for default iterative tracking

### PR DESCRIPTION
**Pull request by mmasciov**
### PR description:
As per title, this PR is introducing a process modifier to enable the usage of parabolic (rather than full) magnetic field for track building, for the default iterative tracking.
It follows a [presentation at the TRK POG](https://indico.cern.ch/event/1247009/#5-selected-topics-in-run-3-tra).

The process modifier is **NOT** enabled by default.
To enable process modifier, the following addition to `cmsDriver.py` command for `step3` is sufficient: 
`--procModifiers=trackingParabolicMf`

For documentation: the usage of the parabolic magnetic field in track building was (unintentionally) reverted in CMSSW 71X, as detailed in issue #41396.

In addition, commit eaf188f2bd1494bac0be78546b237a4e97835d20:
- renames `RecoTracker.TrackProducer.TrackProducerIterativeDefault_cfi.TrackProducer` to `RecoTracker.TrackProducer.TrackProducerIterativeDefault_cfi.TrackProducerIterativeDefault` (as suggested in [comment](https://github.com/cms-sw/cmssw/pull/41557#discussion_r1186550339));
- uses a safer syntax in the corresponding `clone` (as suggested in [comment](https://github.com/cms-sw/cmssw/pull/39578/files#r987113790) from PR #39578).

### PR validation:
As is, this is a **technical PR**: by default, the process modifier is NOT enabled.
To summarize, when the process modifier is enabled: 
- For tracking physics performance: impact on efficiency and fake rate is negligible; impact on resolution is negligible; impact on pT residual at low pT is visible, although small.
- For tracking timing performance: overall track building (full tracking) time is reduced by less than 5% (~2%).

More details can be found in [TRK POG presentation](https://indico.cern.ch/event/1247009/#5-selected-topics-in-run-3-tra), or in the [linked MTV results for TTbar events with PU](http://uaf-10.t2.ucsd.edu/~mmasciov/MIC/ParabolicMfForBuilding/MTV_forDraftPR_ttbarPU/).

FYI, @mmusich, @slava77

### CODE REVIEW

File: RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py [hunk 1]
```diff
@@ -150,14 +153,15 @@
...
-    alwaysUseInvalidHits = True,
-    estimator = 'detachedQuadStepChi2Est',
-    maxDPhiForLooperReconstruction = 2.0,
-    maxPtForLooperReconstruction = 0.7,
-)
+detachedQuadStepTrajectoryBuilder = RecoTracker.CkfPattern.GroupedCkfTrajectoryBuilder_cfi.GroupedCkfTrajectoryBuilder.clone()
+_detachedQuadStepTrajectoryBuilderParabolicMf = RecoTracker.CkfPattern.GroupedCkfTrajectoryBuilder_cfi.GroupedCkfTrajectoryBuilderParabolicMf.clone()
+trackingParabolicMf.toReplaceWith(detachedQuadStepTrajectoryBuilder,_detachedQuadStepTrajectoryBuilderParabolicMf.clone())
+detachedQuadStepTrajectoryBuilder.trajectoryFilter = dict(refToPSet_ = 'detachedQuadStepTrajectoryFilter')
+detachedQuadStepTrajectoryBuilder.maxCand = 3
```
Comment (slava77) [inline RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py:160]: it seems more practical to do similar to fitting and define smth like `GroupedCkfTrajectoryBuilderIterativeDefault_cfi` and simply replace the MF parts there.
The PR line changes will reduce by an order of magnitude

File: RecoTracker/TkNavigation/plugins/SimpleNavigationSchool.cc [hunk 2]
```diff
@@ -372,3 +372,4 @@ void SimpleNavigationSchool::establishInverseRelations() {
 #include "NavigationSchoolFactory.h"
 #include "TrackingTools/DetLayers/interface/NavigationSchool.h"
 DEFINE_EDM_PLUGIN(NavigationSchoolFactory, SimpleNavigationSchool, "SimpleNavigationSchool");
+DEFINE_EDM_PLUGIN(NavigationSchoolFactory, SimpleNavigationSchool, "SimpleNavigationSchoolParabolicMf");
```
[the remark this answers is not in the document]
Comment (makortel) [reply RecoTracker/TkNavigation/plugins/SimpleNavigationSchool.cc:375]: The problem seems to be in `NavigationSchoolESProducer` using the `ComponentName` as both the EventSetup product label
https://github.com/cms-sw/cmssw/blob/40c34ef43cb4775734b445101ec3471a934791f5/RecoTracker/TkNavigation/plugins/NavigationSchoolESProducer.cc#L40-L45
and as the name of the plugin to create with the `NavigationSchoolFactory`
https://github.com/cms-sw/cmssw/blob/40c34ef43cb4775734b445101ec3471a934791f5/RecoTracker/TkNavigation/plugins/NavigationSchoolESProducer.cc#L61-L62

Perhaps a better solution would be allow decoupling the two?

[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/TkNavigation/plugins/SimpleNavigationSchool.cc]: A working attempt in this direction is now there.

File: RecoTracker/TkNavigation/plugins/NavigationSchoolESProducer.cc [hunk 3]
```diff
@@ -58,13 +59,17 @@ NavigationSchoolESProducer::ReturnType NavigationSchoolESProducer::produce(const
   using namespace edm::es;
 
   //get the geometricsearch tracker geometry
+  if (navigationSchoolPluginName_ == "")
+    navigationSchoolPluginName_ = navigationSchoolName_;
+  const std::string pluginName_(navigationSchoolPluginName_);
```
Comment (makortel) [inline RecoTracker/TkNavigation/plugins/NavigationSchoolESProducer.cc:64]: This could be done already in the constructor.

**Comment by slava77** [comment]
I don't see changes in DisplacedGeneralStep_cff.py and DisplacedRegionalStep_cff.py.
I guess we missed them for fitting as well

File: RecoTracker/TkNavigation/plugins/NavigationSchoolESProducer.cc (hunk 3 above)
Comment (mmasciov) [reply RecoTracker/TkNavigation/plugins/NavigationSchoolESProducer.cc]: Done in the last update.

**Comment by slava77** [comment]
@mmasciov 
BTW, I'm not sure if this should be a draft.

If the goal is to use this for some relval/production tests, then it would have to be a normal PR.

**Comment by mmasciov** [comment]
> @mmasciov BTW, I'm not sure if this should be a draft.
> 
> If the goal is to use this for some relval/production tests, then it would have to be a normal PR.

Can turn to a proper PR if the current implementation is satisfactory, no objection on my side

File: RecoTracker/IterativeTracking/python/MixedTripletStep_cff.py [hunk 4]
```diff
@@ -237,21 +240,21 @@
 # Propagator taking into account momentum uncertainty in multiple scattering calculation.
 import TrackingTools.MaterialEffects.MaterialPropagatorParabolicMf_cff
 import TrackingTools.MaterialEffects.MaterialPropagator_cfi
-mixedTripletStepPropagator = TrackingTools.MaterialEffects.MaterialPropagator_cfi.MaterialPropagator.clone(
-#mixedTripletStepPropagator = TrackingTools.MaterialEffects.MaterialPropagatorParabolicMf_cff.MaterialPropagatorParabolicMF.clone(
-    ComponentName = 'mixedTripletStepPropagator',
-    ptMin         = 0.1
-)
+mixedTripletStepPropagator = TrackingTools.MaterialEffects.MaterialPropagator_cfi.MaterialPropagator.clone()
+_mixedTripletStepPropagatorParabolicMf = TrackingTools.MaterialEffects.MaterialPropagatorParabolicMf_cff.MaterialPropagatorParabolicMF.clone()
```
Comment (slava77) [inline RecoTracker/IterativeTracking/python/MixedTripletStep_cff.py:244]: since we are cloning to a local version, I think that the more appropriate (economic?) variant is to just add one line after the old code
```
trackingParabolicMf.toModify(mixedTripletStepPropagator, SimpleMagneticField = 'ParabolicMf')
```

File: RecoTracker/TrackProducer/python/TrackProducerIterativeDefault_cfi.py [hunk 5]
```diff
@@ -1,9 +1,13 @@
 import FWCore.ParameterSet.Config as cms
 import RecoTracker.TrackProducer.TrackProducer_cfi
 
+# for parabolic magnetic field
+from Configuration.ProcessModifiers.trackingParabolicMf_cff import trackingParabolicMf
+
 TrackProducer = RecoTracker.TrackProducer.TrackProducer_cfi.TrackProducer.clone(
```
Comment (slava77) [inline RecoTracker/TrackProducer/python/TrackProducerIterativeDefault_cfi.py:7]:
```suggestion
TrackProducerIterativeDefault = RecoTracker.TrackProducer.TrackProducer_cfi.TrackProducer.clone(
```
should've been done a while back. Reusing the same name is pretty dangerous, considering how often `from xyz import *` happen.

File: RecoTracker/IterativeTracking/python/MixedTripletStep_cff.py (hunk 4 above)
Comment (mmasciov) [reply RecoTracker/IterativeTracking/python/MixedTripletStep_cff.py]: Done in the last update.

File: RecoTracker/TrackProducer/python/TrackProducerIterativeDefault_cfi.py (hunk 5 above)
Comment (mmasciov) [reply RecoTracker/TrackProducer/python/TrackProducerIterativeDefault_cfi.py]: Done in dedicated commit eaf188f2bd1494bac0be78546b237a4e97835d20.

**Comment by mmasciov** [comment]
> > @mmasciov BTW, I'm not sure if this should be a draft.
> > If the goal is to use this for some relval/production tests, then it would have to be a normal PR.
> 
> Can turn to a proper PR if the current implementation is satisfactory, no objection on my side

Have marked as ready for review (@slava77).

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mmasciov** [comment]
> > I don't see changes in DisplacedGeneralStep_cff.py and DisplacedRegionalStep_cff.py. I guess we missed them for fitting as well
> 
> Yes, no change in those so far. Can be propagated if wished; however, those configurations are not run by default, and all these and previous changes only targeted default iterations. Thus, will not do unless requested.

As a change was necessary in `DisplacedRegionalStep_cff.py`, I ended up propagating changes to both.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by clacaputo** [comment]
+reconstruction
